### PR TITLE
Update Packet Broker API and authentication

### PR DESCRIPTION
**Pull request by johanstokking**
#### Summary

This updates the Packet Broker API that's used by the Packet Broker Agent, plus related dependency changes.

Refs https://github.com/packetbroker/router/issues/18
Refs https://github.com/packetbroker/router/issues/19
Refs https://github.com/packetbroker/router/issues/29
Refs https://github.com/packetbroker/router/issues/30
Refs https://github.com/packetbroker/api/issues/21

#### Changes

- Use updated Packet Broker API V3
- Use Packet Broker Routing V1 service
- Use OAuth 2.0 token-based authentication (mTLS is still supported and config is backwards compatible)

#### Testing

Local, needs further testing

##### Regressions

Doesn't work with currently deployed Packet Broker. This may probably need to go in 3.11.

#### Notes for Reviewers

1. ~Check gogoproto Codec usage~
2. Check what needs to be done with the deprecated configuration. We cannot keep this backward compatible as Packet Broker will stop supporting mTLS when the certs are expired

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,8 +13,13 @@ For details about compatibility between different releases, see the **Commitment
 
 ### Changed
 
+- Packet Broker API version to `v3.2.0` and routing API to `v1.0.0`.
```
Comment (johanstokking) [inline CHANGELOG.md]: Needs tags, waiting for https://github.com/packetbroker/go-api/pull/7

File: pkg/packetbrokeragent/agent.go [hunk 2]
```diff
@@ -122,18 +124,58 @@ func New(c *component.Component, conf *Config, opts ...Option) (*Agent, error) {
...
+			conf.OAuth2.TokenURL,
+			conf.OAuth2.ClientID,
+			conf.OAuth2.ClientSecret,
+			[]string{"networks"},
+			conf.Insecure,
+		))
+		if conf.Insecure {
+			authentication[1] = grpc.WithInsecure()
+		} else {
+			authentication[1] = grpc.WithTransportCredentials(credentials.NewClientTLSFromCert(nil, ""))
```
Comment (htdvisser) [inline pkg/packetbrokeragent/agent.go]: Shouldn't we pass the root CAs of the Stack?

**Review by htdvisser — COMMENTED** [review]
Code-owned files LGTM. Just skimmed over the rest.

**Comment by htdvisser** [comment]
> Check what needs to be done with the deprecated configuration. We cannot keep this backward compatible as Packet Broker will stop supporting mTLS when the certs are expired

I guess we should cover things like that in our compatibility commitment. For now we can just deprecate it and write it in CHANGELOG.

**Comment by johanstokking** [comment]
@rvolosatovs any idea what to do with https://github.com/TheThingsNetwork/lorawan-stack/runs/1577511295?

It seems to me that it's related to the fact that the gRPC status that's actually being marshaled here, contains an `Any`, which is marshaled differently.

https://github.com/TheThingsNetwork/lorawan-stack/blob/523990d3f3e903bbec2cd91e9cac52cec8f41481/pkg/errors/grpc.go#L66-L82

**Comment by johanstokking** [comment]
Upgrading `protobuf` to 1.4.x is too risky and doesn't fall in the scope of this PR. Therefore, Packet Broker API is now also being generated with the same protoc image that The Things Stack uses.

The work done one the custom codec, gRPC status, etc is now in a feature branch, see https://github.com/TheThingsNetwork/lorawan-stack/issues/2798#issuecomment-748642624.

File: pkg/packetbrokeragent/agent.go [hunk 3]
```diff
@@ -53,7 +53,16 @@ const (
 // TenantContextFiller fills the parent context based on the tenant ID.
 type TenantContextFiller func(parent context.Context, tenantID string) (context.Context, error)
 
+// TenantExtracter extracts the tenant ID from the context.
+type TenantExtracter func(context context.Context) string
```
Comment (adriansmares) [inline pkg/packetbrokeragent/agent.go]: I think the correct spelling is `Extractor`.

**Review by adriansmares — APPROVED** [review]
LGTM :rocket:

**Review by htdvisser — APPROVED** [review]
Nothing to add. Code-owned files LGTM.

**Comment by johanstokking** [comment]
Merge is blocked on updating the Packet Broker production infrastructure.

**Pull request merged**